# SPDX-License-Identifier: BSD-2-Clause
# Share wide-string compatibility with the ACL, path and string translation units.
--- a/libarchive/archive_entry.c
+++ b/libarchive/archive_entry.c
@@ -121,24 +121,6 @@
 static const char	*ae_strtofflags(const char *stringp, size_t length,
 		    unsigned long *setp, unsigned long *clrp);
 
-#ifndef HAVE_WCSCPY
-static wchar_t * wcscpy(wchar_t *s1, const wchar_t *s2)
-{
-	wchar_t *dest = s1;
-	while ((*s1 = *s2) != L'\0')
-		++s1, ++s2;
-	return dest;
-}
-#endif
-#ifndef HAVE_WCSLEN
-static size_t wcslen(const wchar_t *s)
-{
-	const wchar_t *p = s;
-	while (*p != L'\0')
-		++p;
-	return p - s;
-}
-#endif
 #ifndef HAVE_WMEMCMP
 /* Good enough for simple equality testing, but not for sorting. */
 #define wmemcmp(a,b,i)  memcmp((a), (b), (i) * sizeof(wchar_t))
--- a/libarchive/archive_string.h
+++ b/libarchive/archive_string.h
@@ -45,6 +45,37 @@
 
 #include "archive.h"
 
+#ifndef HAVE_WCSCPY
+static inline wchar_t * wcscpy(wchar_t *s1, const wchar_t *s2)
+{
+	wchar_t *dest = s1;
+	while ((*s1 = *s2) != L'\0')
+		++s1, ++s2;
+	return dest;
+}
+#endif
+#ifndef HAVE_WCSLEN
+static inline size_t wcslen(const wchar_t *s)
+{
+	const wchar_t *p = s;
+	while (*p != L'\0')
+		++p;
+	return p - s;
+}
+#endif
+#ifndef HAVE_WCSCHR
+static inline wchar_t * wcschr(const wchar_t *s, wchar_t c)
+{
+	for (;;) {
+		if (*s == c)
+			return (wchar_t *)s;
+		if (*s == L'\0')
+			return NULL;
+		++s;
+	}
+}
+#endif
+
 /*
  * Basic resizable/reusable string support similar to Java's "StringBuffer."
  *
--- a/libarchive/archive_pathmatch.h
+++ b/libarchive/archive_pathmatch.h
@@ -27,6 +27,8 @@
 #ifndef ARCHIVE_PATHMATCH_H
 #define ARCHIVE_PATHMATCH_H
 
+#include "archive_string.h"
+
 #ifndef __LIBARCHIVE_BUILD
 #ifndef __LIBARCHIVE_TEST
 #error This header is only to be used internally to libarchive.
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1561,6 +1561,7 @@
 CHECK_FUNCTION_EXISTS_GLIBC(utimensat HAVE_UTIMENSAT)
 CHECK_FUNCTION_EXISTS_GLIBC(vfork HAVE_VFORK)
 CHECK_FUNCTION_EXISTS_GLIBC(wcrtomb HAVE_WCRTOMB)
+check_symbol_exists(wcschr wchar.h HAVE_WCSCHR)
 check_symbol_exists(wcscmp wchar.h HAVE_WCSCMP)
 check_symbol_exists(wcscpy wchar.h HAVE_WCSCPY)
 check_symbol_exists(wcslen wchar.h HAVE_WCSLEN)
--- a/build/cmake/config.h.in
+++ b/build/cmake/config.h.in
@@ -1250,6 +1250,9 @@
 #cmakedefine HAVE_WCRTOMB 1
 
+/* Define if wcschr is available. */
+#cmakedefine HAVE_WCSCHR 1
+
 /* Define to 1 if you have the `wcscmp' function. */
 #cmakedefine HAVE_WCSCMP 1
 
 /* Define to 1 if you have the `wcscpy' function. */
--- a/libarchive/archive_platform.h
+++ b/libarchive/archive_platform.h
@@ -90,6 +90,14 @@
  * headers as required.
  */
 
+#include <errno.h>
+#ifndef ENOTSUP
+#define ENOTSUP EOPNOTSUPP
+#endif
+#ifndef EOVERFLOW
+#define EOVERFLOW ERANGE
+#endif
+
 /* Try to get standard C99-style integer type definitions. */
 #if HAVE_INTTYPES_H
 #include <inttypes.h>
--- a/libarchive/archive_platform.h
+++ b/libarchive/archive_platform.h
@@ -126,81 +126,90 @@
 #define	SSIZE_MAX ((ssize_t)(SIZE_MAX >> 1))
 #endif
 #if !HAVE_DECL_UINT32_MAX
 #define	UINT32_MAX (~(uint32_t)0)
 #endif
 #if !HAVE_DECL_INT32_MAX
 #define	INT32_MAX ((int32_t)(UINT32_MAX >> 1))
 #endif
 #if !HAVE_DECL_INT32_MIN
 #define	INT32_MIN ((int32_t)(~INT32_MAX))
 #endif
 #if !HAVE_DECL_UINT64_MAX
 #define	UINT64_MAX (~(uint64_t)0)
 #endif
 #if !HAVE_DECL_INT64_MAX
 #define	INT64_MAX ((int64_t)(UINT64_MAX >> 1))
 #endif
 #if !HAVE_DECL_INT64_MIN
 #define	INT64_MIN ((int64_t)(~INT64_MAX))
 #endif
 #if !HAVE_DECL_UINTMAX_MAX
 #define	UINTMAX_MAX (~(uintmax_t)0)
 #endif
 #if !HAVE_DECL_INTMAX_MAX
 #define	INTMAX_MAX ((intmax_t)(UINTMAX_MAX >> 1))
 #endif
 #if !HAVE_DECL_INTMAX_MIN
 #define	INTMAX_MIN ((intmax_t)(~INTMAX_MAX))
 #endif
 
-/* Some platforms lack the standard PRIxN/PRIdN definitions. */
-#if !HAVE_INTTYPES_H || !defined(PRIx32) || !defined(PRId32)
+/* Some platforms lack the standard PRIxN/PRIdN/PRIuN definitions. */
+#if !HAVE_INTTYPES_H || !defined(PRIx32) || !defined(PRId32) || !defined(PRIu32)
 #ifndef PRIx32
 #if SIZEOF_INT == 4
 #define PRIx32 "x"
 #elif SIZEOF_LONG == 4
 #define PRIx32 "lx"
 #else
 #error No suitable 32-bit unsigned integer type found for this platform
 #endif
 #endif // PRIx32
 #ifndef PRId32
 #if SIZEOF_INT == 4
 #define PRId32 "d"
 #elif SIZEOF_LONG == 4
 #define PRId32 "ld"
 #else
 #error No suitable 32-bit signed integer type found for this platform
 #endif
 #endif // PRId32
-#endif // !HAVE_INTTYPES_H || !defined(PRIx32) || !defined(PRId32)
+#ifndef PRIu32
+#if SIZEOF_INT == 4
+#define PRIu32 "u"
+#elif SIZEOF_LONG == 4
+#define PRIu32 "lu"
+#else
+#error No suitable 32-bit unsigned integer type found for this platform
+#endif
+#endif // PRIu32
+#endif // !HAVE_INTTYPES_H || !defined(PRIx32) || !defined(PRId32) || !defined(PRIu32)
 
 /*
  * If we can't restore metadata using a file descriptor, then
  * for compatibility's sake, close files before trying to restore metadata.
  */
 #if defined(HAVE_FCHMOD) || defined(HAVE_FUTIMES) || defined(HAVE_ACL_SET_FD) || defined(HAVE_ACL_SET_FD_NP) || defined(HAVE_FCHOWN)
 #define	CAN_RESTORE_METADATA_FD
 #endif
 
 /* Set up defaults for internal error codes. */
 #ifndef ARCHIVE_ERRNO_FILE_FORMAT
 #if HAVE_EFTYPE
 #define	ARCHIVE_ERRNO_FILE_FORMAT EFTYPE
 #else
 #if HAVE_EILSEQ
 #define	ARCHIVE_ERRNO_FILE_FORMAT EILSEQ
 #else
 #define	ARCHIVE_ERRNO_FILE_FORMAT EINVAL
 #endif
 #endif
 #endif
 
 #ifndef ARCHIVE_ERRNO_PROGRAMMER
 #define	ARCHIVE_ERRNO_PROGRAMMER EINVAL
 #endif
 
 #ifndef ARCHIVE_ERRNO_MISC
 #define	ARCHIVE_ERRNO_MISC (-1)
 #endif
 
--- a/libarchive/archive_time.c
+++ b/libarchive/archive_time.c
@@ -115,44 +115,42 @@
 	if (dt > DOS_MAX_TIME) {
 		dt = DOS_MAX_TIME;
 	}
 	else if (dt < DOS_MIN_TIME) {
 		dt = DOS_MIN_TIME;
 	}
 	return dt;
 }
 
 /* Convert NTFS time to Unix sec/nsec */
 void
 ntfs_to_unix(uint64_t ntfs, int64_t* secs, uint32_t* nsecs)
 {
 	if (ntfs > INT64_MAX) {
 		ntfs -= NTFS_EPOC_TICKS;
 		*secs = ntfs / NTFS_TICKS;
 		*nsecs = 100 * (ntfs % NTFS_TICKS);
 	}
 	else {
-		lldiv_t tdiv;
 		int64_t value = (int64_t)ntfs - (int64_t)NTFS_EPOC_TICKS;
 
-		tdiv = lldiv(value, NTFS_TICKS);
-		*secs = tdiv.quot;
-		*nsecs = (uint32_t)(tdiv.rem * 100);
+		*secs = value / (int64_t)NTFS_TICKS;
+		*nsecs = (uint32_t)((value % (int64_t)NTFS_TICKS) * 100);
 	}
 }
 
 /* Convert Unix sec/nsec to NTFS time */
 uint64_t
 unix_to_ntfs(int64_t secs, uint32_t nsecs)
 {
 	uint64_t ntfs;
 
 	if (secs < -(int64_t)NTFS_EPOC_TIME)
 		return 0;
 
 	ntfs = secs + NTFS_EPOC_TIME;
 
 	if (ntfs > UINT64_MAX / NTFS_TICKS)
 		return UINT64_MAX;
 
 	ntfs *= NTFS_TICKS;
 
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1545,6 +1545,7 @@
 CHECK_FUNCTION_EXISTS_GLIBC(strchr HAVE_STRCHR)
 CHECK_FUNCTION_EXISTS_GLIBC(strdup HAVE_STRDUP)
 CHECK_FUNCTION_EXISTS_GLIBC(strerror HAVE_STRERROR)
+check_symbol_exists(strtoimax "inttypes.h" HAVE_STRTOIMAX)
 CHECK_FUNCTION_EXISTS_GLIBC(strncpy_s HAVE_STRNCPY_S)
 CHECK_FUNCTION_EXISTS_GLIBC(strnlen HAVE_STRNLEN)
 CHECK_FUNCTION_EXISTS_GLIBC(strrchr HAVE_STRRCHR)
--- a/build/cmake/config.h.in
+++ b/build/cmake/config.h.in
@@ -1049,6 +1049,9 @@
 /* Define to 1 if you have the `strerror_r' function. */
 #cmakedefine HAVE_STRERROR_R 1
 
+/* Define to 1 if strtoimax is available. */
+#cmakedefine HAVE_STRTOIMAX 1
+
 /* Define to 1 if you have the `strftime' function. */
 #cmakedefine HAVE_STRFTIME 1
 
--- a/libarchive/archive_platform.h
+++ b/libarchive/archive_platform.h
@@ -98,6 +98,12 @@
 #include <stdint.h>
 #endif
 
+#ifndef HAVE_STRTOIMAX
+#include <stdlib.h>
+_Static_assert(sizeof(intmax_t) == sizeof(long long), "strtoll must cover intmax_t");
+#define strtoimax strtoll
+#endif
+
 /* Borland warns about its own constants!  */
 #if defined(__BORLANDC__)
 # if HAVE_DECL_UINT64_MAX
